src/Parser/CkanResourceUrlParser.php

```diff
@@ -47,20 +47,36 @@ public function __construct(CkanClientInterface $ckan_client, LoggerInterface $l
    * {@inheritdoc}
    */
   public function parse($url) {
-    $options = parse_url($url);
-    $path = !empty($options['path']) ? $options['path'] : '';
-
-    if (!preg_match(self::CKAN_RESOURCE_PATH_REGEX, $path)) {
+    if (!preg_match(self::CKAN_RESOURCE_PATH_REGEX, $url)) {
       return FALSE;
     }
 
-    // Expected format /dataset/PACKAGE_ID/resource/RESOURCE_ID.
-    $parts = explode('/', $options['path'], 5);
-    $options = [
-      'package_id' => $parts[2],
-      'resource_id' => $parts[4],
-    ];
+    /**
+     * There are three types of url
+     * /dataset/.../resource/RESOURCE_ID
+     * /dataset/PACKAGE_ID/resource/RESOURCE_ID/download/...
+     * /data/api/.../action/datastore_search?resource_id=RESOURCE_ID...
+     */
+    $options = parse_url($url);
+    if(!empty($options['query'])) {
+      parse_str($options['query'], $query_string);
+    }
 
+    //Get the resource_id from the query string or parse the url to get the resource_id
```

Comment:
As per my comment on your other PR, comments should always have a space after `//` 
Comments should always end with a full stop too

https://www.drupal.org/node/1354#inline